File: backend/internal/network/traffic.go
package network

import (
	"fmt"
)

type TrafficManager struct {
	runner Runner
}

func NewTrafficManager(runner Runner) *TrafficManager {
	return &TrafficManager{runner: runner}
}

func (m *TrafficManager) SetupRoot(iface string) error {
	_, err := m.runner.Execute("tc", "qdisc", "add", "dev", iface, "root", "handle", "1:", "htb", "default", "10")
	return err
}

func (m *TrafficManager) LimitClient(iface, ip string, rateKbps int) error {
	// Example: tc class add dev wlan0 parent 1: classid 1:1 htb rate 1024kbps
	// tc filter add dev wlan0 protocol ip parent 1:0 prio 1 u32 match ip dst 192.168.1.5 flowid 1:1
	
	classID := "1:100" // Should be dynamic based on IP or a map
	_, err := m.runner.Execute("tc", "class", "add", "dev", iface, "parent", "1:", "classid", classID, "htb", "rate", fmt.Sprintf("%dkbit", rateKbps))
	if err != nil {
		return err
	}

	_, err = m.runner.Execute("tc", "filter", "add", "dev", iface, "protocol", "ip", "parent", "1:0", "prio", "1", "u32", "match", "ip", "dst", ip, "flowid", classID)
	return err
}

func (m *TrafficManager) Clear(iface string) error {
	_, err := m.runner.Execute("tc", "qdisc", "del", "dev", iface, "root")
	return err
}

File: backend/app/core/traffic.py
import logging
from ..utils.system import run_command

logger = logging.getLogger(__name__)

async def initialize_tc(interface: str = "wlan0") -> None:
    """Initialize tc qdisc root on the interface"""
    # clear existing qdisc
    await run_command(["sudo", "tc", "qdisc", "del", "dev", interface, "root"])
    
    # set up root qdisc
    code, _, err = await run_command([
        "sudo", "tc", "qdisc", "add", "dev", interface, "root", "handle", "1:", "htb", "default", "10"
    ])
    if code != 0 and "File exists" not in err:
        logger.error(f"Failed to initialize tc: {err}")

async def apply_bandwidth_limit(interface: str, ip_address: str, class_id: int, rate_kbit: int) -> None:
    """Apply an HTB bandwidth limit to an IP address"""
    hex_id = f"1:{class_id}"
    
    # Create the class for the specific rate
    await run_command([
        "sudo", "tc", "class", "replace", "dev", interface, "parent", "1:", "classid", hex_id,
        "htb", "rate", f"{rate_kbit}kbit"
    ])
    
    # Add sfq to ensure fairness within the class
    await run_command([
        "sudo", "tc", "qdisc", "replace", "dev", interface, "parent", hex_id, "handle", f"{class_id}:", "sfq", "perturb", "10"
    ])
    
    # Filter traffic destined to the IP into the new class
    await run_command([
        "sudo", "tc", "filter", "replace", "dev", interface, "protocol", "ip", "parent", "1:", "prio", "1",
        "u32", "match", "ip", "dst", ip_address, "flowid", hex_id
    ])

async def remove_bandwidth_limit(interface: str, class_id: int) -> None:
    """Remove a specific HTB class"""
    hex_id = f"1:{class_id}"
    await run_command([
        "sudo", "tc", "class", "del", "dev", interface, "classid", hex_id
    ])
